File: src/Array/ProductOfArrayExceptSelf/ProductOfArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        
        //Get prefix
        res[0] = 1;
        for(int i = 1; i < res.length; i++){
            res[i] = res[i - 1] * nums[i - 1];
        }
        
        //Get product array except self
        int postFix = 1;
        for(int i = res.length - 1; i >= 0; i--){
            res[i] *= postFix;
            postFix *= nums[i];
        }
        return res;
    }
}

File: src/Array/ProductOfArrayExceptSelf/ProductOfArrayExceptSelf.py
from typing import *


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # 1. Calculate the prefix
        res = [1]
        for i in range(1, len(nums)):
            res.append(res[i - 1] * nums[i - 1])
        
        # 2. use the array to calculate the product except self
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            prefix = res[i]
            res[i] = postfix * prefix 
            postfix *= nums[i]
        
        return res
